File: examples/cpp/example02.cc
#include <iostream>

struct Volume { int ml; };
struct Moisture { int percentage; };
struct Temperature { double celsius; };

class ThermoSensor {
public:
  bool read(Temperature & temperature) {
    temperature.celsius = 21.5;
    // return false;
    return true;
  }
};

class MoistureSensor {
public:
  bool read(Moisture & moisture) {
    moisture.percentage = 40;
    // return false;
    return true;
  }
};

class Pump {
public:
  bool pump(Volume amount) {
    // return false;
    return true;
  }
};

class WateringSystem {
public:
  bool water(Volume & amount) {
    Moisture moisture;
    if(moisture_sensor.read(moisture)) {
      Temperature temperature;
      if(thermo_sensor.read(temperature)) {
        if(calculate_amount(moisture, temperature, amount)) {
          if(pump.pump(amount)) {
            return true;
          }
        }
      }
    }
    return false;
  }

private:
  bool calculate_amount(Moisture moisture, Temperature temperature, Volume & amount) {
    amount.ml = 178;
    return true;
  }

  MoistureSensor moisture_sensor;
  ThermoSensor thermo_sensor;
  Pump pump;
};

int main() {
  Volume amount;
  if(WateringSystem{}.water(amount)) {
    std::cout << "Water " << amount.ml << " ml\n";
  } else {
    std::cout << "Watering error\n";
  }
}

File: examples/rust/example02.rs
struct Moisture {
    percentage: i32,
}

struct Temperature {
    celsius: f64,
}

struct Volume {
    ml: i32,
}

fn read_moisture() -> Result<Moisture, String> {
    // Err(String::from("Moisture sensor error"))
    Ok(Moisture{percentage: 40})
}

fn read_temperature() -> Result<Temperature, String> {
    // Err(String::from("Temperature sensor error"))
    Ok(Temperature{celsius: 21.5})
}

fn calculate_amount(m: Moisture, t: Temperature) -> Result<Volume, String> {
    Ok(Volume{ml: 178})
}

fn pump(v: &Volume) -> Result<(), String> {
    // Err(String::from("Pump error"))
    Ok(())
}

fn water_plant() -> Result<Volume, String> {
    let moisture = read_moisture()?;
    let temperature = read_temperature()?;
    let amount = calculate_amount(moisture, temperature)?;
    pump(&amount)?;
    Ok(amount)
}

fn main() {
    match water_plant() {
        Err(e) => println!("{}", e),
        Ok(amount) => println!("Water {:?} ml", amount.ml),
    }
}
